src/pyats/contrib/creators/topology.py

```diff
@@ -0,0 +1,775 @@
+        """
+        # create 3 sets to track what devices have cdp and lldp configured
+        # and what devices have be visited by the script
+        self.cdp_configured = set()
+        self.lldp_configured = set()
+        self.visited_devices = set()
+        self.ad = {}
+        return {
+            'required': ['testbed_name'],
+            'optional': {
+                'config': False,
+                'all_interfaces': False,
+                'exclude_network': '',
+                'exclude_interfaces':'',
+                'topo_only': False,
+                'timeout': 10,
+                'alias': ''
+            }
+        }
+    
+    def connect_one_device(self, testbed, device):
+        '''
+            connect to the given device in the testbed using the given
+            connections and after that enable cdp and lldp if allowed
+        '''
+        # if the device has a supported os try to connect through each of 
+        # the connections it has. If the device is not connected just return
+        # that cdp and lldp were not configured
+        if testbed.devices[device].os not in supported_os:
+            return [False, False]
+
+        # if there is a prefered alias for the device, attempt to connect with device
+        # using that alias, if the attmept fails or the alias doesn't exist, it will
+        # attempt to connect normally
+        if device in self.ad:
+            if self.ad[device] in testbed.devices[device].connections:
+                log.info('Attempting to connect to {} with alias {}'.format(device, self.ad[device]))
+                try:
+                    testbed.devices[device].connect(via = str(self.ad[device]),
+                                                    connection_timeout = 10,
+                                                    learn_hostname = True)
+                except:        
+                    log.info('Failed to connect to {} with alias {}'.format(device, self.ad[device]))
+                    testbed.devices[device].destroy()
+            else:
+                log.info('Device {} does not have a connection with alias {}'.format(device, self.ad[device]))
+
+        if testbed.devices[device].is_connected():
+            return self.configure_device_cdp_and_lldp(testbed.devices[device])
+
+        for one_connect in testbed.devices[device].connections:                         
+            try:
+                testbed.devices[device].connect(via = str(one_connect),
+                                                connection_timeout = 10,
+                                                learn_hostname = True)
+                break
+            except Exception:
+                # if connection fails, erase the connection from connection mgr
+                testbed.devices[device].destroy()
+
+        return self.configure_device_cdp_and_lldp(testbed.devices[device])
+        
+    def configure_device_cdp_and_lldp(self, dev):

```

Comment:
`configure_discovery_protocols`? I'm concerning when we need to add different protocols in the future. better not to have protocol names in func name.

Reply:
Done